python/ql/src/experimental/semmle/python/frameworks/Django.qll

```diff
@@ -70,6 +70,301 @@ private module Django {
+          module RawSQL {
+            /** Gets a reference to the `django.db.models.expressions.RawSQL` class. */
+            private DataFlow::Node classRef(DataFlow::TypeTracker t) {
+              t.start() and
+              result = DataFlow::importNode("django.db.models.expressions.RawSQL")
+              or
+              t.start() and
+              result = DataFlow::importNode("django.db.models.RawSQL") // Commonly used alias
+              or
+              t.startInAttr("RawSQL") and
+              result = expressions()
+              or
+              exists(DataFlow::TypeTracker t2 | result = classRef(t2).track(t2, t))
+            }
+
+            /**
+             * Gets a reference to the `django.db.models.expressions.RawSQL` class.
+             */
+            DataFlow::Node classRef() { result = classRef(DataFlow::TypeTracker::end()) }
+
+            /** Gets an instance of the `django.db.models.expressions.RawSQL` class. */
+            private DataFlow::Node classInstance(DataFlow::TypeTracker t, ControlFlowNode sql) {
+              t.start() and
+              exists(CallNode c | result.asCfgNode() = c |
+                c.getFunction() = classRef().asCfgNode() and
+                c.getArg(0) = sql
+              )
+              or
+              exists(DataFlow::TypeTracker t2 | result = classInstance(t2, sql).track(t2, t))
+            }
+
+            /** Gets an instance of the `django.db.models.expressions.RawSQL` class. */
+            DataFlow::Node classInstance(ControlFlowNode sql) {
+              result = classInstance(DataFlow::TypeTracker::end(), sql)
+            }
+          }
+        }
+      }
+    }
+
+    /**
+     * A call to the `django.db.connection.cursor.execute` function.
+     *
+     * See
+     * - https://docs.djangoproject.com/en/3.1/topics/db/sql/#executing-custom-sql-directly
+     * - https://docs.djangoproject.com/en/3.1/topics/db/sql/#connections-and-cursors
+     */
+    private class DbConnectionExecute extends SqlExecution::Range, DataFlow::CfgNode {
+      override CallNode node;
+
+      DbConnectionExecute() { node.getFunction() = django::db::execute().asCfgNode() }
+
+      override DataFlow::Node getSql() {
+        result.asCfgNode() in [node.getArg(0), node.getArgByName("sql")]
+      }
+    }
+
+    /**
+     * A call to the `annotate` function on a model using a `RawSQL` argument.
+     *
+     * See https://docs.djangoproject.com/en/3.1/ref/models/querysets/#annotate
+     */
+    private class ObjectsAnnotate extends SqlExecution::Range, DataFlow::CfgNode {
+      override CallNode node;
+      ControlFlowNode sql;
+
+      ObjectsAnnotate() {
+        node.getFunction() = django::db::models::objects_attr("annotate").asCfgNode() and
+        django::db::models::expressions::RawSQL::classInstance(sql).asCfgNode() in [node.getArg(_),
+              node.getArgByName(_)]
+      }
+
+      override DataFlow::Node getSql() { result.asCfgNode() = sql }
+    }
```

Comment:
I think in general, this is the wrong way to handle taint-propagation, and instead we should make the result of `RawSQL(tainted)`  also be tainted (or maybe even data-flow preserving).

With the current setup, if you have code like this (which might not make any sense)

```py
# foo.py

from sql_helper import sql_helper

def handle_http_request(request):
    raw_sql = RawSQL("select age from user where username = '" + request.GET["username"] + "'")
    sql_result = sql_helper(raw_sql)
    ...

# sql_helper.py

def sql_helper(raw_sql):
    return User.objects.annotate(raw_sql)
```

We would generate an alert that goes from the `request` parameter in `handle_http_request`, to the construction of `RawSQL` -- but it wouldn't be obvious from the alert _where_ the SQL is actually executed :open_mouth: 

Personally I don't know if handling this correctly will ever give us interesting results, so let us just record that this bit could be more polished, and move on to more important things :wink:

Reply:
Would we? The sink is the call to `annotate`. Although I do agree on the principle of preferring taint/data-flow over type tracking.

Reply:
Well the sink ends up being the result of `getSql` on a `SqlExecution::Range` right?